content/blog/15-making-rust-attractive-for-writing-gtk-applications.md

```diff
@@ -0,0 +1,301 @@
++++
+title = "Making Rust attractive for writing GTK applications"
+date = 2022-10-23
+draft = false
+
+[taxonomies]
+categories = ["Rust"]
+tags = ["Rust"]
+
+[extra]
+lang = "en"
+toc = true
+math = false
+mermaid = false
+cc_license = true
++++
+
+[Rust](https://www.rust-lang.org/), the programming language, have been gaining traction across the whole stack. From the pretty recent [upstream kernel support](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=8aebac82933ff1a7c8eede18cab11e1115e2062b), writing games, websites, low level OS components or even desktop applications.
+
+The gtk-rs team have been doing an impressive amount of work during the last few years to make the experience of using `GObject` based libraries from Rust enjoyable by providing high quality bindings around those libraries, generated with [gir](https://github.com/gtk-rs/gir) from the [introspection data](https://gi.readthedocs.io/en/latest/).
+
+Approximately two years ago and few months before the release of [GTK](https://gtk.org/) 4, I decided to take over the maintenance of [gtk4-rs](https://github.com/gtk-rs/gtk4-rs) and push forward the initial work made by Xiang Fan during a Google Summer of Code internship. Nowadays, it is the most used GTK 4 bindings out there with probably more than 100 applications written in it, from simple applications like [Contrast](https://flathub.org/apps/details/org.gnome.design.Contrast) to complex ones like [Telegrand](https://github.com/melix99/telegrand) or [Spot](https://flathub.org/apps/details/dev.alextren.Spot).
+
+In the following post, I will be talking about the current status and what we have achieved since the first release of `gtk4-rs`.
+
+## Bindings
+
+As mentioned above, a good portion of the bindings are generated automatically using [gir](https://github.com/gtk-rs/gir) but in some cases a manual implementation is needed like in the following cases:
+
+- Make the code more Rust idiomatic
+- Handling cases that are too specific to be supported by [gir](https://github.com/gtk-rs/gir)
+- Writing the necessary infrastructre code to create custom GObjects and so allowing the end user to make their own custom widgets or GStreamer plugins for example
+
+TODO: add some numbers about the amount of manual code in gtk4-rs
+
+### Subclassing
+
+In the early days of `gtk3-rs`, the infrastructure for writing custom subclassses wasn't fully there yet, especially that the amount of manual code that has to be written for supporting all the virtual functions (that can be overridden by a sub-implementation) of `gtk::Widget` was a lot. That caused people to avoid writing custom GTK widgets and do plenty of hacks like having one single `ObjectWrapper` GObject that would serialize a Rust struct into a `JSON` to store it as a string property in order to store these Rust types in a `gio::ListModel` and use it with `gtk::ListBox::bind_model`/`gtk::FlowBox::bind_model`.
```

Comment:
```suggestion
In the early days of `gtk3-rs`, the infrastructure for writing custom subclassses wasn't fully there yet, especially the amount of manual code that had to be written for supporting all the virtual functions (that can be overridden by a sub-implementation) of `gtk::Widget` was a lot. That caused people to avoid writing custom GTK widgets and do plenty of hacks like having one single `ObjectWrapper` GObject that would serialize a Rust struct into a `JSON` to store it as a string property in order to store these Rust types in a `gio::ListModel` and use it with `gtk::ListBox::bind_model`/`gtk::FlowBox::bind_model`.
```